Framework/DataObjects/test/ScanningWorkspaceBuilderTest.h

```diff
@@ -0,0 +1,509 @@
+      for (size_t j = 0; j < nDetectors; ++j) {
+        const auto index = i * nDetectors + j;
+        TS_ASSERT_EQUALS(spectrumDefinitions[index].size(), 1)
+        TS_ASSERT_EQUALS(spectrumDefinitions[index][0].first, j)
+        TS_ASSERT_EQUALS(spectrumDefinitions[index][0].second, i)
+        TS_ASSERT_EQUALS(detectorIDs[spectrumDefinitions[index][0].first],
+                         j + 1)
+      }
+    }
+  }
+
+  void test_setting_indexing_type_twice_throws_and_error() {
+    const auto &instrument = createSimpleInstrument(nDetectors, nBins);
+
+    auto builder = ScanningWorkspaceBuilder(nDetectors, nTimeIndexes, nBins);
+    TS_ASSERT_THROWS_NOTHING(builder.setInstrument(instrument));
+    TS_ASSERT_THROWS_NOTHING(builder.setTimeRanges(timeRanges));
+    TS_ASSERT_THROWS_NOTHING(builder.setIndexingType(
+        ScanningWorkspaceBuilder::IndexingType::DETECTOR_ORIENTED))
+    TS_ASSERT_THROWS_EQUALS(
+        builder.setIndexingType(
+            ScanningWorkspaceBuilder::IndexingType::TIME_ORIENTED),
+        const std::logic_error &e, std::string(e.what()),
+        "Indexing type has been set already.")
+  }
+
+private:
+  size_t nDetectors = 5;
+  size_t nTimeIndexes = 4;
+  size_t nBins = 10;
+
+  std::vector<std::pair<DateAndTime, DateAndTime>> timeRanges = {
+      {0, 1}, {1, 3}, {3, 6}, {6, 10}};
+
+  std::vector<double> timeDurations = {1e-9, 2e-9, 3e-9, 4e-9};
+
+  std::vector<std::vector<V3D>> positions;
+  std::vector<std::vector<Quat>> rotations;
+  std::vector<double> instrumentAngles;
+
+  void initalisePositions(size_t nDetectors, size_t nTimeIndexes) {
+    for (size_t i = 0; i < nDetectors; ++i) {
+      std::vector<V3D> timePositions;
+      for (size_t j = 0; j < nTimeIndexes; ++j) {
+        timePositions.push_back(V3D(double(i), double(j), 1.0));
+      }
+      positions.push_back(timePositions);
+    }
+  }
+
+  void initaliseRotations(size_t nDetectors, size_t nTimeIndexes) {
+    for (size_t i = 0; i < nDetectors; ++i) {
+      std::vector<Quat> timeRotations;
+      for (size_t j = 0; j < nTimeIndexes; ++j) {
+        timeRotations.push_back(Quat(double(i), double(j), 1.0, 2.0));
+      }
+      rotations.push_back(timeRotations);
+    }
+  }
+
+  void initialiseInstrumentAngles(size_t nTimeIndexes) {
+    for (size_t i = 0; i < nTimeIndexes; ++i) {
+      instrumentAngles.push_back(double(i) * 30.0);
+    }
+  }
+
+  Instrument_const_sptr createSimpleInstrument(size_t nDetectors,
+                                               size_t nBins) {
+    const auto &wsWithInstrument =
+        WorkspaceCreationHelper::create2DWorkspaceWithFullInstrument(
+            int(nDetectors), int(nBins));
+    return wsWithInstrument->getInstrument();
+  }
+
+  void checkTimeRanges(const DetectorInfo &detectorInfo) {
+    for (size_t i = 0; i < nDetectors; ++i) {
+      for (size_t j = 0; j < nTimeIndexes; ++j) {
+        TS_ASSERT_EQUALS(detectorInfo.scanInterval({i, j}), timeRanges[j]);
+      }
+    }
+  }
+};
+
```

Comment:
No Performance test. I think tracking performance would probably be a good idea.